### DISCUSSION TITLE
[@types/testing-library__jest-dom] incompatible with `expect` from `@jest/globals`

### DISCUSSION THREAD

**Discussion by elado** [DefinitelyTyped/DefinitelyTyped#55362] [Issues with a @types package]
Types from `@types/testing-library__jest-dom` for `@testing-library/jest-dom` are incompatible with `expect` from `@jest/globals` because they extend `@types/jest` types

Works:

```tsx
import { render } from '@testing-library/react';

it('renders text in child prop', () => {
  const div = render(<div>Hello</div>);
  expect(div).toHaveTextContent('Hello');
});
```

Doesn't work:

```tsx
import { it, expect } from '@jest/globals';
import { render } from '@testing-library/react';

it('renders text in child prop', () => {
  const div = render(<div>Hello</div>);
  expect(div).toHaveTextContent('Hello');
});
```

Error:
```
Property 'toHaveTextContent' does not exist on type 'Matchers<RenderResult<typeof import(".../node_modules/@testing-library/dom/types/queries"), HTMLElement>>'.ts(2339)
```

Global types from `@types/jest` are problematic and conflicting with other libraries (chai, cypress) and importing directly from `@jest/globals` helps with that.

**The root cause:**

- `@jest/globals` imports `expect` directly from `expect` library with its own `Matchers` interface
- `@types/jest` has a local version of `expect` with a different `Matchers` interface
- RTL extends `declare namespace jest { interface Matchers ..` which is the one exclusively in `@types/jest` and not the one in `@jest/globals`

**Potential solutions:** 

- Have `@types/jest` use the original `expect` types instead of the local ones
- Have `@types/testing-library__jest-dom` extend also original `expect` types for `Matchers`

[a comment by an automated account is not shown]

**Reply by jgoz** [reply]
It's been a while since I looked at this, but is it possible for a DT package to augment types from an outside module? Both of your suggestions would require the ability to augment types in `@jest/globals` from somewhere in `@types`. That probably means there needs to be an npm dependency on the former from the latter, and while it may work on a technical level, I'm not sure if it's a supported use case in DT.

**Reply by elado** [reply]
Looks like @SimenB tried to merge `@types/jest` and `@jest/globals` which would help https://github.com/DefinitelyTyped/DefinitelyTyped/pull/44365 but that work stalled.

As for augmenting `expect` as well: 
I'm not sure it's possible, as it's not a shared/global namespace like the one in `@types/jest`, but I could be wrong https://unpkg.com/browse/expect@27.1.1/build/index.d.ts

**Reply by SimenB** [reply]
#44365 is the solution for `@types/jest` having duplicated types, but e.g. `@types/testing-library__jest-dom` would still only mess with a global `@types/jest` I think. Unless interface merging would just work :tm:
